alibi_detect/ad/adversarialae.py

```diff
@@ -139,7 +139,7 @@ def fit(self,
             loss_fn: tf.keras.losses = loss_adv_ae,
             w_model: float = 1.,
             w_recon: float = 0.,
-            optimizer: tf.keras.optimizers = tf.keras.optimizers.Adam(learning_rate=1e-3),
+            optimizer: tf.keras.optimizers.Optimizer = tf.keras.optimizers.Adam,
```

Comment:
Avoid instantiation in default kwarg, since this meant `optimizer` was instantiated upon detector import instead of instantiation. See [PR discussion](https://github.com/SeldonIO/alibi-detect/pull/739#:~:text=to%20ClassifierDriftTF.-,Outlier%20and%20adversarial%20detectors,-For%20most%20of).

Reply:
Note: the type signature needs to be updated here, but we need to discuss the best way to do this since it could get messy. Technically the type is now `Union[Type[tf.keras.optimizers.Optimizer], tf.keras.optimizers.Optimizer, Type[tf.keras.optimizers.legacy.Optimizer], tf.keras.optimizers.legacy.Optimizer]`

Reply:
We should be getting some `mypy` errors from tests in that case? (in fact the old type doesn't make sense...)

Reply:
Yeh the old types don't make sense, but this is a wider issue with typing of  `kernel`, `optimizer`, `dataset` at maybe other kwargs, so think it makes sense to leave the fix to a follow-up>

Reply:
p.p.s on second thoughts, the `optimizer` typing issue is introduced by this PR, so we should fix it here...

The question is how to do this, the new `Union` would be quite ugly, so maybe we should declare the `Union` in `alibi_detect.utils._types` instead?

Reply:
Yea, I would agree with having a well-named type alias.

Reply:
Since `tensorflow` is optional, there might be some issue with having these types transparently available...

Reply:
> Since `tensorflow` is optional, there might be some issue with having these types transparently available...

I've added a type alias in [53bd5f1](https://github.com/SeldonIO/alibi-detect/pull/739/commits/53bd5f1472e266d349f3caaf89b321cc38624a68), with postponed annotations so things still work when `tensorflow` isn't installed. Is this what you're referring to?